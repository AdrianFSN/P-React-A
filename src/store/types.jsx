// auth state types
export const AUTH_LOGIN = "auth/login";

export const AUTH_LOGIN_PENDING = "auth/login/pending";
export const AUTH_LOGIN_FULFILLED = "auth/login/fulfilled";
export const AUTH_LOGIN_REJECTED = "auth/login/rejected";

export const AUTH_LOGOUT = "auth/logout";

// advertisements state types
export const ADS_LOADED = "ads/loaded";

export const ADS_LOADED_PENDING = "ads/loaded/pending";
export const ADS_LOADED_FULFILLED = "ads/loaded/fulfilled";
export const ADS_LOADED_REJECTED = "ads/loaded/rejected";

export const ADS_CREATED = "ads/created";
export const ADS_CREATED_PENDING = "ads/created/pending";
export const ADS_CREATED_FULFILLED = "ads/created/fulfilled";
export const ADS_CREATED_REJECTED = "ads/created/rejected";

export const ADS_DETAIL = "ads/detail";
export const ADS_DETAIL_PENDING = "ads/detail/pending";
export const ADS_DETAIL_FULFILLED = "ads/detail/fulfilled";
export const ADS_DETAIL_REJECTED = "ads/detail/rejected";

export const ADS_DELETED_PENDING = "ads/deleted/pending";
export const ADS_DELETED_FULFILLED = "ads/deleted/fulfilled";
export const ADS_DELETED_REJECTED = "ads/deleted/rejected";

//errors related types
export const UI_RESET_ERROR = "ui/reset-error";
