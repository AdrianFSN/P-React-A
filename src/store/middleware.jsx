export const failureRedirects =
  (router, redirectsMap) => (store) => (next) => (action) => {
    const result = next(action);

    if (!action.error) {
      return result;
    }

    const redirect = redirectsMap[action.payload.response.status];
    if (redirect) {
      router.navigate(redirect);
    }

    return result;
  };
