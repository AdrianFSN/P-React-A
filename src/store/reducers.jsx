import {
  ADS_LOADED_FULFILLED,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
  ADS_DETAIL_FULFILLED,
  ADS_CREATED_FULFILLED,
  ADS_DELETED_FULFILLED,
  ADS_REQUEST_DELETION,
  ADS_CONFIRM_DELETION,
  ADS_CANCEL_DELETION,
  ADS_TAGS_FULFILLED,
  ADS_MAX_PRICE,
  ADS_MIN_PRICE,
} from "./types";

const defaultState = {
  auth: false,
  ads: {
    loaded: false,
    data: [],
    availableTags: [],
    deletionRequest: false,
    confirmDeletion: false,
    prices: {
      maxPriceAvailable: 0,
      minPriceAvailable: 0,
    },
  },
  ui: {
    pending: false,
    error: null,
  },
};

// auth related reducer
export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_FULFILLED:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function ads(state = defaultState.ads, action) {
  switch (action.type) {
    case ADS_LOADED_FULFILLED:
      return { ...state, loaded: true, data: action.payload };
    case ADS_CREATED_FULFILLED:
      return { ...state, data: [action.payload, ...state.data] };
    case ADS_TAGS_FULFILLED:
      return { ...state, availableTags: action.payload };
    case ADS_DETAIL_FULFILLED:
      return { ...state, data: [action.payload] };
    case ADS_DELETED_FULFILLED:
      return {
        ...state,
        data: state.data.filter((ad) => ad.id !== action.payload.id),
        deletionRequest: false,
        confirmDeletion: false,
      };
    case ADS_REQUEST_DELETION:
      return { ...state, deletionRequest: true };
    case ADS_CONFIRM_DELETION:
      return { ...state, confirmDeletion: true };
    case ADS_CANCEL_DELETION:
      return { ...state, deletionRequest: false, confirmDeletion: false };
    default:
      return state;
  }
}

export function prices(state = defaultState.ads.prices, action) {
  switch (action.type) {
    case ADS_MAX_PRICE:
      return {
        ...state,
        maxPriceAvailable: action.payload,
      };
    case ADS_MIN_PRICE:
      return {
        ...state,
        minPriceAvailable: action.payload,
      };
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { ...state, pending: false, error: action.payload };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  if (action.type.endsWith("/pending")) {
    return { ...state, pending: true };
  }

  if (action.type.endsWith("/fulfilled")) {
    return { ...state, pending: false, error: null };
  }

  return state;
}
