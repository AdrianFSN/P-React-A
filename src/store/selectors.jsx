export const getIsLogged = (state) => state.auth;

export const areAdvertsLoaded = (state) => state.ads.loaded;
export const getListOfAds = (state) => state.ads.data;
/* export const getAdDetail = (advertId) => (state) => {
  return getListOfAds(state).find((ad) => ad.id === Number(advertId));
}; */

export const getAdDetail = (state, advertId) => {
  getListOfAds(state).find((ad) => ad.id === Number(advertId));
};

export const getUi = (state) => state.ui;
