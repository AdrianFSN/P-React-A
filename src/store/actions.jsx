//import * as advertsService from "../pages/adverts/service";
//import { getAdvert, getLatestAds } from "../pages/adverts/service";
//import { login } from "../pages/auth/service";
import { areAdvertsLoaded, getAdDetail } from "./selectors";
import {
  ADS_LOADED_FULFILLED,
  ADS_LOADED_PENDING,
  //ADS_CREATED,
  ADS_CREATED_FULFILLED,
  ADS_CREATED_PENDING,
  ADS_CREATED_REJECTED,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_REJECTED,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
  ADS_DETAIL_PENDING,
  ADS_DETAIL_FULFILLED,
  ADS_DETAIL_REJECTED,
} from "./types";

// actions related to auth state
export const authLogin = (credentials, storageRequest) => {
  return async function (dispatch, _getState, { services: { auth }, router }) {
    try {
      dispatch(authLoginPending());
      await auth.login(credentials, storageRequest);
      dispatch(authLoginFulfilled());
      const to = router.state.location.state?.from || "/";
      router.navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginRejected(error));
      throw error;
    }
  };
};

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const authLoginPending = () => ({
  type: AUTH_LOGIN_PENDING,
});

export const authLoginFulfilled = () => ({
  type: AUTH_LOGIN_FULFILLED,
});

export const authLoginRejected = (error) => ({
  type: AUTH_LOGIN_REJECTED,
  payload: error,
  error: true,
});

export const adsLoadedPending = () => ({
  type: ADS_LOADED_PENDING,
});

export const adsLoadedFulfilled = (ads) => ({
  type: ADS_LOADED_FULFILLED,
  payload: ads,
});

export const adsLoadedRejected = (error) => ({
  type: ADS_LOADED_FULFILLED,
  payload: error,
  error: true,
});

export const loadAdverts = () => {
  return async function (dispatch, getState, { services }) {
    const state = getState();
    if (areAdvertsLoaded(state)) {
      return;
    }

    try {
      dispatch(adsLoadedPending());
      const adverts = await services.ads.getLatestAds();
      dispatch(adsLoadedFulfilled(adverts));
    } catch (error) {
      dispatch(adsLoadedRejected(error));
    }
  };
};

export const adsCreatedPending = () => ({
  type: ADS_CREATED_PENDING,
});
export const adsCreatedFulfilled = (ad) => ({
  type: ADS_CREATED_FULFILLED,
  payload: ad,
});
export const adsCreatedRejected = (error) => ({
  type: ADS_CREATED_REJECTED,
  payload: error,
  error: true,
});

export const adsDetailPending = () => ({
  type: ADS_DETAIL_PENDING,
});
export const adsDetailFulfilled = (ad) => ({
  type: ADS_DETAIL_FULFILLED,
  payload: ad,
});
export const adsDetailRejected = (error) => ({
  type: ADS_DETAIL_REJECTED,
  payload: error,
  error: true,
});

export const loadAdvert = (advertId) => {
  return async function (dispatch, getState, { services, router }) {
    const state = getState();
    if (getAdDetail(advertId)(state)) {
      return;
    }
    try {
      dispatch(adsDetailPending);
      const advert = await services.ads.getAdvert(advertId);
      dispatch(adsDetailFulfilled(advert));
      router.navigate(`/adverts/${advert.id}`);
    } catch (error) {
      dispatch(adsDetailRejected(error));
      throw error;
    }
  };
};

export const createAdvert = (advert) => {
  return async function (dispatch, _getService, { services, router }) {
    try {
      dispatch(adsCreatedPending());
      const { id } = await services.ads.createNewAd(advert);
      const createdAdvert = await services.ads.getAdvert(id);
      dispatch(adsCreatedFulfilled(createdAdvert));
      router.navigate(`/adverts/${createdAdvert.id}`);
      return createdAdvert;
    } catch (error) {
      dispatch(adsCreatedRejected(error));
    }
  };
};

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
