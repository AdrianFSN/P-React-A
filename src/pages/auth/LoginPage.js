import { useState } from "react";
import { login } from "./service";
import Button from "../../components/shared/Button";
import FormField from "../../components/shared/FormField";
import CheckBox from "../../components/shared/CheckBox";
import "./LoginPage.css";
import {
  authLogin,
  authLoginFulfilled,
  authLoginPending,
  authLoginRejected,
  uiResetError,
} from "../../store/actions";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../store/selectors";
import { useLocation, useNavigate } from "react-router-dom";

export default function LoginPage() {
  const dispatch = useDispatch();
  const location = useLocation();
  const navigate = useNavigate();
  const [formValues, setFormValues] = useState({
    email: "",
    password: "",
  });
  const { pending, error } = useSelector(getUi);

  const [checkBoxStatus, setCheckBoxStatus] = useState(false);

  const handleCheckboxChange = () => {
    setCheckBoxStatus((prevStatus) => !prevStatus);
  };

  const handleChange = (event) => {
    setFormValues((currentFormValues) => ({
      ...currentFormValues,
      [event.target.name]: event.target.value,
    }));
  };
  const { email, password } = formValues;
  const buttonDisabled = !email || !password || pending;

  const handleSubmit = async (event) => {
    event.preventDefault();

    try {
      dispatch(authLoginPending());
      await login(formValues, checkBoxStatus);
      dispatch(authLoginFulfilled());
      dispatch(authLogin());

      const to = location.state?.from || "/";
      navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginRejected(error));
    }
  };
  const resetError = () => {
    dispatch(uiResetError());
  };

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to Nodepop</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          className="loginForm-field"
          type="text"
          name="email"
          value={email}
          onChange={handleChange}
          placeholder="Your email here"
        ></FormField>
        <FormField
          className="loginForm-field"
          type="password"
          name="password"
          value={password}
          onChange={handleChange}
          placeholder="Your password here"
        ></FormField>
        <Button
          className="loginForm-submit"
          type="submit"
          $variant="primary"
          disabled={buttonDisabled}
        >
          Login
        </Button>
        <CheckBox
          label="Click here to keep your session open"
          checked={checkBoxStatus}
          onChange={handleCheckboxChange}
        ></CheckBox>
      </form>
      <div>
        {error && (
          <div
            className="Nodepop-error"
            onClick={resetError}
          >{`${error}. Click this banner to get back`}</div>
        )}
      </div>
    </div>
  );
}
