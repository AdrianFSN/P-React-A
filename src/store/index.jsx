import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import * as reducers from "./reducers";
import * as actionCreators from "./actions";
import * as auth from "../pages/auth/service";
import * as ads from "../pages/adverts/service";
import { withExtraArgument } from "redux-thunk";
import { failureRedirects } from "./middleware";

const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({ actionCreators });

export default function configureStore(preloadedState, { router }) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(
      applyMiddleware(
        withExtraArgument({ services: { auth, ads }, router }),
        failureRedirects(router, {
          401: "/login",
          404: "/404",
        })
      )
    )
  );

  return store;
}
