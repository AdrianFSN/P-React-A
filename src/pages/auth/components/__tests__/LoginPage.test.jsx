import { fireEvent, render, screen } from "@testing-library/react";
import LoginPage from "../../LoginPage";
import { Provider } from "react-redux";
import { authLogin } from "../../../../store/actions";
import { toBeChecked } from "@testing-library/jest-dom/matchers";

jest.mock("../../../../store/actions");

describe("LoginPage", () => {
  const state = {
    ui: {
      pending: false,
      error: null,
    },
  };

  const store = {
    dispatch: () => {},
    getState: () => state,
    subscribe: () => {},
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

  it("should take a snapshot for LoginPage Component", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("should dispatch authLogin action", () => {
    const email = "username@example.com";
    const password = "password";
    const checkBoxStatus = true;
    renderComponent();
    const emailInput = screen.getByPlaceholderText(/Your email here/i);
    /* const emailInput = screen.getByRole("textbox", {
      name: "",
    }); */
    const passwordInput = screen.getByPlaceholderText(/Your password here/i);
    const submitButton = screen.getByRole("button");
    const checkbox = screen.getByRole("checkbox");

    expect(submitButton).toBeDisabled();

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    fireEvent.click(checkbox);

    expect(checkbox).toBeChecked();

    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email, password }, checkBoxStatus);
  });
});
