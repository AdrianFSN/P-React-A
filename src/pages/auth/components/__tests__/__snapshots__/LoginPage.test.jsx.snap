// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should take a snapshot for LoginPage Component 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Nodepop
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span />
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            placeholder="Your email here"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span />
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            placeholder="Your password here"
            type="password"
            value=""
          />
        </label>
      </div>
      <button
        class="sc-beySPh hXHSwk loginForm-submit"
        disabled=""
        type="submit"
      >
        Login
      </button>
      <div
        class="checkBox"
      >
        <label
          class="checkBox-label"
        >
          <input
            class="checkBox-input"
            type="checkbox"
          />
          <span>
            Click here to keep your session open
          </span>
        </label>
      </div>
    </form>
    <div />
  </div>
</div>
`;

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Nodepop
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span />
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            placeholder="Your email here"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span />
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            placeholder="Your password here"
            type="password"
            value=""
          />
        </label>
      </div>
      <button
        class="sc-beySPh hXHSwk loginForm-submit"
        disabled=""
        type="submit"
      >
        Login
      </button>
      <div
        class="checkBox"
      >
        <label
          class="checkBox-label"
        >
          <input
            class="checkBox-input"
            type="checkbox"
          />
          <span>
            Click here to keep your session open
          </span>
        </label>
      </div>
    </form>
    <div />
  </div>
</div>
`;
